File: discord/src/main/java/io/github/shaksternano/borgar/discord/util/Fonts.java
package io.github.shaksternano.borgar.discord.util;

import io.github.shaksternano.borgar.discord.Main;
import io.github.shaksternano.borgar.discord.io.FileUtil;

import java.awt.*;
import java.io.IOException;

/**
 * Handles custom fonts.
 */
public class Fonts {

    /**
     * Registers the custom fonts.
     */
    public static void registerFonts() {
        FileUtil.forEachResource("font", (resourcePath, inputStream) -> {
            try {
                Font font = Font.createFont(Font.TRUETYPE_FONT, inputStream);
                GraphicsEnvironment.getLocalGraphicsEnvironment().registerFont(font);
            } catch (FontFormatException | IOException e) {
                Main.getLogger().error("Could not load font " + resourcePath, e);
            }
        });
    }

    public static boolean fontExists(String fontName) {
        for (var font : GraphicsEnvironment.getLocalGraphicsEnvironment().getAllFonts()) {
            if (font.getName().equals(fontName)) {
                return true;
            }
        }
        return false;
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/graphics/Fonts.kt
package io.github.shaksternano.borgar.core.graphics

import io.github.shaksternano.borgar.core.io.forEachResource
import io.github.shaksternano.borgar.core.logger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.awt.Font
import java.awt.GraphicsEnvironment
import java.io.InputStream

const val DEFAULT_FONT_NAME = Font.DIALOG

suspend fun registerFonts() = forEachResource(
    "font"
) { resourcePath: String, inputStream: InputStream ->
    runCatching {
        val font = withContext(Dispatchers.IO) {
            Font.createFont(Font.TRUETYPE_FONT, inputStream)
        }
        GraphicsEnvironment.getLocalGraphicsEnvironment().registerFont(font)
    }.onFailure {
        logger.error("Error loading font $resourcePath", it)
    }
}

fun fontExists(fontName: String): Boolean =
    GraphicsEnvironment.getLocalGraphicsEnvironment().allFonts.any {
        it.name == fontName
    }
